### ISSUE TITLE
Interest in supporting masking?

### ISSUE THREAD

**Issue by mortonjt** [cvxgrp/cvxpy#411]
First off, this is a really awesome interface.  Really simplifies prototyping convex optimizations.

One thing that isn't clear at the moment (if it is possible) is how to explicitly mask entries within a matrix during optimization.  Here is an example using ALS with matrix completion

```python
def ALS(R, U, M):
    """
    Updates the factors using Alternating Least Squares.

    Performs alternating least squares.  Specifically, it will try
    to solve the following problem

    min_UV ||R - UV'||^2

    Parameters
    ----------
    R : scipy.sparse.matrix
        N x D matrix of observed ratings.
    U : np.array
        N x r latent factor matrix of rank r.
    M : scipy.sparse.matrix
        Mask for indicating indices of observed values

    Returns
    -------
    U : np.array
        Updated latent factor matrix of N x r
    V : np.array
        Updated latent factor matrix of D x r
    """
    N, r = U.shape
    N, D = R.shape
    Ux = Variable(N, r)
    Vy = Variable(r, D)
    Mc = Constant(M)
    Rc = Constant(R)

    objective = Minimize(
        sum_entries(
            square(
                mul_elemwise(Mc, U * Vy - Rc)
            )
        )
    )
    prob = Problem(objective)
    prob.solve()
    V = Vy.value

    objective = Minimize(
        sum_entries(
            square(
                mul_elemwise(Mc, Ux * V - Rc)
            )
        )
    )
    prob = Problem(objective)
    prob.solve()
    U = Ux.value

    return U, V.T
```

Here, I'm specifying two convex optimization problems, to solve the matrix completion problem.  Note that this does work nicely on small datasets.  However, this becomes problematic with large sparse datasets.  Its mainly because I don't want to evaluate every entry in the `U * Vy` or the `Ux * V` entries, because the resulting dense matrix sizes could blow up the memory if `N` and `D` are large.  

1. Is it possible to enable explicit masking within the  `U * Vy` or the `Ux * V` matrices so that not all of the dot products are evaluated, and thus a sparse matrix is returned
2. If not, is there interest for supporting this sort of operation in the future?

**Comment by SteveDiamond** [comment]
Sorry, I don't quite understand your question and what you mean by masking. The point is you want sparse variables? I'm planning to add that feature in cvxpy 1.0.

**Comment by garnoli99** [comment]
Hi, appreciate this is a very old post but I have a similar need and cannot find any other posts on masking for cvxpy. 

I would like to mask a 2d variable so that I only create constraints for some of elements in the variable matrix. I noticed I can do this using a 1d mask perfectly fine, but using a 2d masking approach similar to what numpy offers via `np.ma.MaskedArray` does not work.

For example, this 1d array approach works fine:
```
x = cvx.Variable(shape=(3,3))
mask = np.array([True, True, False]) 

x[:, mask] >= some_parameter
```

...but for a 2d array using `np.ma.MaskedArray` to retain the shape, it does not:
```
x = cvx.Variable(shape=(3,3))
mask = np.array([[True, True, False], [True, False, False], [True, True, True]]) 

x[mask] >= some_parameter
```
In this example I would like the constraints to apply to all elements in the Variable matrix apart from elements (1,3), (2,2) and (2,3). However I get `{MaskError}Mask and data not compatible: data size is 1, mask size is 9`.

One solution could be to just multiply the mask matrix by `some_parameter` in my constraint:
```
x = cvx.Variable(shape=(3,3))
mask = np.array([[True, True, False], [True, False, False], [True, True, True]]) 

x >= mask * some_parameter
```

But Im wondering if this would have the exact same performance as masking the Variable array. If so I am more than happy to use this work around!

Any thoughts would be very much appreciated, thanks!

**Comment by SteveDiamond** [comment]
What version of cvxpy are you using? When I run
```
x = cvx.Variable(shape=(3,3))
mask = np.array([[True, True, False], [True, False, False], [True, True, True]]) 
```
``x[mask]`` is a totally valid cvxpy expression, which selects the ``True`` entries.

**Comment by garnoli99** [comment]
Thanks for your response, and sorry I realized my example is incorrect (just edited it now). I also should have been clearer: `x[mask]` does work but it results the matrix being flattened (`shape=(6,)`) which was causing issues elsewhere in my formulation. 

To retain a shape of (3,3) I tried to use `np.ma.MaskedArray(data=x, mask=mask)` and this gives the error `{MaskError}Mask and data not compatible: data size is 1, mask size is 9`.

Note: I have now adopted the multiplication approach to mask the matrix (the last example in my post above) and this works perfectly fine. So I no longer need the masking approach, I just wanted to round off the discussion.
